src/sage/schemes/elliptic_curves/hom_sum.py

```diff
@@ -0,0 +1,644 @@
+
+        EXAMPLES::
+
+            sage: E = EllipticCurve(GF(101), [5,5])
+            sage: phi = E.isogenies_prime_degree(7)[0]
+            sage: (phi + phi)._degree_bound()
+            28
+            sage: (phi + phi).degree()
+            28
+            sage: (phi - phi)._degree_bound()
+            28
+            sage: (phi - phi).degree()
+            0
+
+        ::
+
+            sage: E = EllipticCurve(GF(443), [1,1])
+            sage: pi = E.frobenius_endomorphism()
+            sage: m1 = E.scalar_multiplication(1)
+            sage: (pi - m1)._degree_bound()
+            486
+            sage: (pi - m1).degree()
+            433
+        """
+        B = ZZ(0)
+        for phi in self._phis:
+            B += phi.degree() + 2*sqrt(B * phi.degree())
+        return integer_floor(B)
+
+    def _compute_degree(self):
+        r"""
+        Internal method to compute and cache the degree of this sum morphism
+        (and its dual).
+
+        ALGORITHM: Evaluate the composition with the dual on points of small
+        order and solve logarithms to eventually recover the degree using CRT.
+        (This is essentially Schoof's algorithm, applied to a scalar.)
+
+        EXAMPLES::
+
+            sage: E = EllipticCurve(GF(101), [5,5])
+            sage: phi = E.isogenies_prime_degree(7)[0]
+            sage: isog = phi + phi
+            sage: print(isog._degree)
+            None
+            sage: isog._compute_degree()
+            sage: isog._degree
+            28
+
+        ::
+
+            sage: E = EllipticCurve(GF(443), [1,1])
+            sage: pi = E.frobenius_endomorphism()
+            sage: m1 = E.scalar_multiplication(1)
+            sage: endo = pi - m1
+            sage: print(endo._degree)
+            None
+            sage: endo._compute_degree()
+            sage: endo._degree
+            433
+            sage: endo.dual()._degree
+            433
+        """
+        if self._degree is not None:
+            return
+        if len(self._phis) == 0:
+            self._degree = 0
+        elif len(self._phis) == 1:
+            self._degree = self._phis[0].degree()
+        else:
+            #TODO In some cases it would probably be faster to simply
+            # compute the kernel polynomial using the addition formulas?
+            from sage.rings.finite_rings.integer_mod import Mod
+
+            M = self._degree_bound() + 1
+            deg = Mod(0,1)
+            for l in Primes():
+                if deg.modulus() >= M:
+                    break
+                try:
+                    P = point_of_order(self._domain, l)
+                except ValueError:
+                    continue   # supersingular and l == p
+
+                Q = self.dual()._eval(self._eval(P))
+                d = discrete_log(Q, P, ord=l, operation='+')
+                deg = deg.crt(Mod(d, l))
+
+            self._degree = ZZ(deg)
+            self.dual()._degree = self._degree
```

Comment:
Does this line really do anything?  Doesn't it create the dual, set its degree, and then throw it away?

Reply:
In this class `.dual()` is marked `@cached_method`, so the same object will be returned on subsequent calls.